plugins/field-dependent-dropdown/src/field_dependent_dropdown.ts

```diff
@@ -0,0 +1,274 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview A Blockly dropdown menu field where the options can change
+ * depending on the value of another field.
+ */
+
+import * as Blockly from 'blockly/core';
+import {DependentDropdownOptionsChange} from './dependent_dropdown_options_change';
+
+/** The type of the mapping from parent value to child options. */
+export interface ChildOptionMapping {
+  [key: string]: Blockly.MenuOption[];
+}
+
+// This type isn't exported from Blockly so we have to derive it from the API.
+type FieldConfig =
+    Exclude<ConstructorParameters<typeof Blockly.Field>[2], undefined>;
+
+/** fromJson config for a dependent dropdown field. */
+export interface FieldDependentDropdownFromJsonConfig extends FieldConfig {
+  parentName: string;
+  optionMapping: ChildOptionMapping;
+}
+
+/**
+ * A structure for managing data needed by the menu generator of a
+ * FieldDependentDropdown.
+ */
+interface DependencyData {
+  /**
+   * A reference to the parent field of an associated dependent dropdown. Absent
+   * until the parent field is initialized. This field's value determines the
+   * available options of the child field.
+   */
+  parentField?: Blockly.Field<string>;
+
+  /**
+   * The child field's currently available menu options based on the current
+   * value of the parent field. Absent until the parent field is initialized.
+   */
+  derivedOptions?: Blockly.MenuOption[];
+}
+
+/**
+ * A dropdown field that automatically updates its own options based on a
+ * mapping from a parent field's value. The parent field must be attached to the
+ * block before this child field.
+ *
+ * When this field is attached to a block, it will find the parent field and
+ * attach a validator to the parent field that intercepts changes to its value
+ * to update the options on this field. If the new validator is later removed\
+ * or replaced, then this dependent field will no longer function.
+ */
+export class FieldDependentDropdown extends Blockly.FieldDropdown {
```

Comment:
As a follow up feature request (you don't have to do as part of this PR) it would be great if you could configure a default set of options to use if the parent field's value is not in the optionMapping. That would enable this to more easily be used with parent fields that aren't necessarily dropdown fields.

Reply:
Good idea, I went ahead and implemented it here, thanks!